feat(kafka sink): Add support for compression

bruceg opened the pull request:
Closes #1963

Commit:
Move the common `fn to_string` into the serde module

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Add `Default` implementation for the kafka config structs

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Add support for compression to the kafka source and sink config

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Merge common kafka sink and source option documentation

Signed-off-by: Bruce Guenter <bruce@timber.io>

binarylogic commented on .meta/_partials/_kafka_common.toml:
```diff
@@ -0,0 +1,43 @@
+[<%= namespace %>.options.compression]
+type = "string"
+common = true
+examples = ["none","gzip"]
```
You don't need this if you provide an `enum`.

bruceg replied:
Fixed

binarylogic reviewed (approved):
Looks good!

Commit:
Drop the redundant "examples" for the compression enum

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Drop the `pub(crate)` from to_string to satisfy the component checker

When all components are disabled, it produces a warning that
`crate::serde::to_string` is never used.

Signed-off-by: Bruce Guenter <bruce@timber.io>

a-rodin commented:
Is it possible to add testing of all of these compression options to the integration tests of the `kafka` sink?

Commit:
Add support for zstd to the kafka library

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Add tests for compression modes to the kafka sink

Signed-off-by: Bruce Guenter <bruce@timber.io>

bruceg commented:
I've added a separate test for each of the compression options. This required setting the `zstd` option on the `rdkafka` crate, which pulls in the `zstd-sys` crate unconditionally.

a-rodin commented on Cargo.toml:
```diff
@@ -113,7 +113,7 @@ derive_is_enum_variant = "0.1.1"
 db-key = "0.0.5"
 headers = "0.2.1"
-rdkafka = { version = "0.23.1", features = ["ssl"], optional = true }
+rdkafka = { version = "0.23.1", features = ["ssl","zstd"], optional = true }
```
Maybe can also add `libz` feature here to be explicit?

a-rodin commented:
This looks good, I think it can be merged.

Not sure what to do with the failing semantic PR check. From [conventionalcommits.org](https://www.conventionalcommits.org/en/v1.0.0/#what-do-i-do-if-the-commit-conforms-to-more-than-one-of-the-commit-types):
> > What do I do if the commit conforms to more than one of the commit types?
>
> Go back and make multiple commits whenever possible. Part of the benefit of Conventional Commits is its ability to drive us to make more organized commits and PRs.

As I understand it, if we are following Conventional Commits, it is probably better to split this into two separate PRs, one for the source and the other for the sink. As an alternative we may just remove the scope and keep only the `feat:` prefix. I'm curious what does @binarylogic think about this.

binarylogic commented:
> it is probably better to split this into two separate PRs

That's correct, but we can allow for multiple scopes in our release process. So I'm indifferent. In general, though, we should keep change small and focused. I don't think this violates that.

Commit:
Add explicit feature flag for zlib use in the rdkafka crate

Signed-off-by: Bruce Guenter <bruce@timber.io>

lukesteensen commented on src/sinks/kafka.rs:
```diff
@@ -97,6 +101,10 @@ impl KafkaSinkConfig {
         }
+        client_config.set(
+            "compression.codec",
+            &to_string(self.compression.unwrap_or(KafkaCompression::default())),
```
I think we could use `unwrap_or_default()` here.

bruceg replied:
Of course.

lukesteensen commented on src/sources/kafka.rs:
```diff
@@ -160,6 +162,10 @@ fn create_consumer(config: KafkaSourceConfig) -> crate::Result<StreamConsumer> {
         )
         .set("enable.auto.offset.store", "false")
+        .set(
+            "compression.codec",
```
This is only a producer-side configuration item. The compression format of a record batch is encoded at write time and the consumer will check the header and decompress accordingly.

bruceg replied:
Hmm, ok, the original issue asked for support in both the sink and source. Reading [this Kafka document](https://cwiki.apache.org/confluence/display/KAFKA/Compression) clarifies that this indeed is a producer-only. Is the `compression.topics` configuration setting also required to make this work?

lukesteensen replied:
I don't think we need to be concerned with the topic-level configuration option. That's more for management of the kafka cluster itself.

I believe the `Not implemented` error we saw was just due to not having the relevant features enabled on the crate, so it did not have an available implementation of the decompression algorithm for that record batch. There is not setting to tell a consumer how to decompress a batch because that information is encoded as part of the wire protocol.

lukesteensen reviewed (changes requested):
Sink side looks good! I believe all we should need to do for the source/consumer side is enable the relevant features of `librdkafka`.

Commit:
Use `.unwrap_or_default`

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Drop the compression setting from the source, not relevant there

Signed-off-by: Bruce Guenter <bruce@timber.io>

binarylogic commented:
@bruceg once the conflict is fixed let's merge.

Commit:
Merge remote-tracking branch 'origin/master' into kafka-compression

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Merge remote-tracking branch 'origin/master' into kafka-compression

Signed-off-by: Bruce Guenter <bruce@timber.io>
